lib/astutils.cpp

```diff
@@ -3525,3 +3525,51 @@ bool isSizeOfEtc(const Token *tok)
 {
     return Token::Match(tok, "sizeof|typeof|offsetof|decltype|__typeof__ (");
 }
+
+std::string extractVariableName(const Variable *var)
+{
+    if (!var)
+        return "";
+
+    return formatTypeName(var->typeStartToken(), var->typeEndToken()->next());
+}
+
+std::string formatTypeName(const Token *start, const Token *end)
+{
+    if (!start || !end)
+        return "";
+
+    std::string name;
+
+    // single token type?
+    if (start->next() == end) {
+        name = start->str();
+    }
+
+    // complicated type
+    else {
+        const Token *tok = start;
+        int level = 0;
+
+        while (tok && tok != end) {
```

Comment:
I am confused.. I have the feeling that the end token should be included in the returned string also.. it should normally point at the last type token.

Reply:
The standard in C++ iterators (and these start/finish are used in the same manner) is that the end points to one-after the last element, which means that the end should not be used.
Also, if we take a look at `findTypeEnd` for example, it uses the same convention.